XenModel/Utils/Helpers.cs

```diff
@@ -959,31 +959,120 @@ public static void RemoveFromGuiConfig(Session session, IXenObject o, string key
 
         #endregion
 
+        public enum DataSourceCategory
+        {
+            Cpu,
+            Memory,
+            Disk,
+            Storage,
+            Network,
+            Latency,
+            LoadAverage,
+            Gpu,
+            Pvs,
+            Custom
+        }
+
+        public static string ToStringI18N(this DataSourceCategory category)
+        {
+            switch (category)
+            {
+                case DataSourceCategory.Cpu:
+                    return Messages.DATATYPE_CPU;
+                case DataSourceCategory.Memory:
+                    return Messages.DATATYPE_MEMORY;
+                case DataSourceCategory.Disk:
+                    return Messages.DATATYPE_DISK;
+                case DataSourceCategory.Storage:
+                    return Messages.DATATYPE_STORAGE;
+                case DataSourceCategory.Network:
+                    return Messages.DATATYPE_NETWORK;
+                case DataSourceCategory.Latency:
+                    return Messages.DATATYPE_LATENCY;
+                case DataSourceCategory.LoadAverage:
+                    return Messages.DATATYPE_LOADAVERAGE;
+                case DataSourceCategory.Gpu:
+                    return Messages.DATATYPE_GPU;
+                case DataSourceCategory.Pvs:
+                    return Messages.DATATYPE_PVS;
+                default:
+                    return Messages.DATATYPE_CUSTOM;
+            }
+        }
+
         public static Regex CpuRegex = new Regex("^cpu([0-9]+)$");
-        public static Regex CpuAvgFreqRegex = new Regex("^CPU([0-9]+)-avg-freq$");
+        static Regex CpuAvgFreqRegex = new Regex("^CPU([0-9]+)-avg-freq$");

```

Comment:
This is the only regex with uppercase lettering. So just double checking that's on purpose here.

Reply:
Yes, there are some host datasources that use uppercase lettering. You can open a shell to the host and check the names on the CLI: `xe host-data-source-list uuid=xxxx | grep name_label` (and similarly for a running VM with `vm-data-source-list`).